IPC Testing API: asynchronous stream messages with replies not working

kkinnunen-apple opened the pull request:
#### 08baee0f736008bb945716a2430aa9a513040f36
<pre>
IPC Testing API: asynchronous stream messages with replies not working
<a href="https://bugs.webkit.org/show_bug.cgi?id=272148">https://bugs.webkit.org/show_bug.cgi?id=272148</a>
<a href="https://rdar.apple.com/125084127">rdar://125084127</a>

Reviewed by Ryosuke Niwa.

Implement support for asynchronous replies in JS testing API:
 - JSIPCConnection::sendWithAsyncReply
 - JSIPCConnection::waitForAsyncReplyAndDispatchImmediately
 - JSIPCStreamClientConnection::sendWithAsyncReply
 - JSIPCStreamClientConnection::waitForAsyncReplyAndDispatchImmediately

Previously in JS, the JSIPC object would support sending the async message
and receiving it as a promise. Stream connections would not support it.

Replace it with more consistent connection-specific sendWithAsyncReply
calls that take a callback parameter, similar to C++. Implement the
promise variant in JS side.

This way the async reply id is available for the tests, if they need it.

Organize the code so that the JSIPCConnection and
JSIPCStreamClientConnection share the common implementation in jsSend,
jsSendSync, jsSendWithAsyncReply, jsWaitForMessage,
jsWaitForAsyncReplyAndDispatchImmediately. Organize
JSIPCStreamClientConnection::prepareToSendOutOfStreamMessage to just
prepare the stream connection so that the above common functionality can
then be shared.

* LayoutTests/resources/ipc.js:
(connectionSendWithPromisedReply):
(IPC.sendWithPromisedReply):
(asyncFlush):
* Source/WebKit/Shared/IPCStreamTester.cpp:
(WebKit::IPCStreamTester::asyncPing):
(WebKit::IPCStreamTester::asyncMessage): Deleted.
* Source/WebKit/Shared/IPCStreamTester.h:
* Source/WebKit/Shared/IPCStreamTester.messages.in:
* Source/WebKit/Shared/IPCTester.cpp:
(WebKit::IPCTester::asyncPing):
* Source/WebKit/Shared/IPCTester.h:
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::JSIPCConnection::create):
(WebKit::IPCTestingAPI::JSIPCConnection::connection const):
(WebKit::IPCTestingAPI::JSIPCConnection::JSIPCConnection):
(WebKit::IPCTestingAPI::createErrorFromIPCError):
(WebKit::IPCTestingAPI::extractSyncIPCMessageInfo):
(WebKit::IPCTestingAPI::jsSend):
(WebKit::IPCTestingAPI::jsSendWithAsyncReply):
(WebKit::IPCTestingAPI::jsSendSync):
(WebKit::IPCTestingAPI::jsWaitForMessage):
(WebKit::IPCTestingAPI::jsWaitForAsyncReplyAndDispatchImmediately):
(WebKit::IPCTestingAPI::JSIPCConnection::staticFunctions):
(WebKit::IPCTestingAPI::JSIPCConnection::sendMessage):
(WebKit::IPCTestingAPI::JSIPCConnection::sendWithAsyncReply):
(WebKit::IPCTestingAPI::JSIPCConnection::sendSyncMessage):
(WebKit::IPCTestingAPI::JSIPCConnection::waitForMessage):
(WebKit::IPCTestingAPI::JSIPCConnection::waitForAsyncReplyAndDispatchImmediately):
(WebKit::IPCTestingAPI::JSIPCStreamClientConnection::staticFunctions):
(WebKit::IPCTestingAPI::extractIPCStreamMessageInfo):
(WebKit::IPCTestingAPI::JSIPCStreamClientConnection::prepareToSendOutOfStreamMessage):
(WebKit::IPCTestingAPI::JSIPCStreamClientConnection::sendMessage):
(WebKit::IPCTestingAPI::JSIPCStreamClientConnection::sendWithAsyncReply):
(WebKit::IPCTestingAPI::JSIPCStreamClientConnection::sendSyncMessage):
(WebKit::IPCTestingAPI::JSIPCStreamClientConnection::waitForMessage):
(WebKit::IPCTestingAPI::JSIPCStreamClientConnection::waitForAsyncReplyAndDispatchImmediately):
(WebKit::IPCTestingAPI::JSIPC::staticFunctions):
(WebKit::IPCTestingAPI::JSIPC::processTargetFromArgument):
(WebKit::IPCTestingAPI::JSIPC::addMessageListener):
(WebKit::IPCTestingAPI::messageNameFromArgument):
(WebKit::IPCTestingAPI::encodeArgument):
(WebKit::IPCTestingAPI::JSIPC::getConnectionForProcessTarget):
(WebKit::IPCTestingAPI::JSIPC::sendMessage):
(WebKit::IPCTestingAPI::JSIPC::waitForMessage):
(WebKit::IPCTestingAPI::JSIPC::sendSyncMessage):
(WebKit::IPCTestingAPI::sendMessageWithJSArguments): Deleted.
(WebKit::IPCTestingAPI::sendSyncMessageWithJSArguments): Deleted.
(WebKit::IPCTestingAPI::waitForMessageWithJSArguments): Deleted.
(WebKit::IPCTestingAPI::processTargetFromArgument): Deleted.
(WebKit::IPCTestingAPI::messageIDFromArgument): Deleted.
* LayoutTests/ipc/async-with-reply-expected.txt: Added.
* LayoutTests/ipc/async-with-reply.html: Added.
* LayoutTests/ipc/stream-async-with-reply-expected.txt: Added.
* LayoutTests/ipc/stream-async-with-reply.html: Added.

Canonical link: <a href="https://commits.webkit.org/277239@main">https://commits.webkit.org/277239@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8c3270df424526a2c50b493b154f941278079125

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47105 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26280 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49744 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49788 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43154 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31536 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23741 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38370 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47686 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/23756 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40595 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19677 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/21159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41745 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5149 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/43483 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42149 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51663 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22129 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/18507 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45661 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23406 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/40741 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44665 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10390 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24188 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23123 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on LayoutTests/resources/ipc.js:
```diff
@@ -2,10 +2,31 @@ function randomIPCID() {
+    }
+}
+
 function asyncFlush(processTarget) {
```
Can we make this function async and use `await` in the function instead of manually calling then?

rniwa commented on LayoutTests/resources/ipc.js:
```diff
@@ -2,10 +2,31 @@ function randomIPCID() {
     return Math.floor(Math.random() * 10000) + 1;
 }
 
+function connectionSendWithPromisedReply(connection, destinationID, messageName, args) {
```
I would have called this sendWithPromisedReplyForConnection or something.
It's odd / sounds unnatural to prefix the function name with "connection".

rniwa commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -2380,61 +2487,96 @@ static JSC::JSObject* jsResultFromReplyDecoder(JSC::JSGlobalObject* globalObject
 {
-    RefPtr jsIPC = toWrapped(context, thisObject);
-    if (!jsIPC) {
+    RefPtr self = toWrapped(context, thisObject);
```
I wouldn't call this "self". Maybe "impl" but I think I would have preferred to keep the old name: jsIPC.

rniwa commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -2380,61 +2487,96 @@ static JSC::JSObject* jsResultFromReplyDecoder(JSC::JSGlobalObject* globalObject
 {
-    RefPtr jsIPC = toWrapped(context, thisObject);
-    if (!jsIPC) {
+    RefPtr self = toWrapped(context, thisObject);
```
I don't think we should do this variable rename.

rniwa commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -2380,61 +2487,96 @@ static JSC::JSObject* jsResultFromReplyDecoder(JSC::JSGlobalObject* globalObject
+
+JSValueRef JSIPC::sendMessage(JSContextRef context, JSObjectRef, JSObjectRef thisObject, size_t argumentCount, const JSValueRef arguments[], JSValueRef* exception)
+{
+    RefPtr self = toWrapped(context, thisObject);
```
I don't think we should do this variable rename.

rniwa commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -2380,61 +2487,96 @@ static JSC::JSObject* jsResultFromReplyDecoder(JSC::JSGlobalObject* globalObject
-JSValueRef JSIPC::sendMessage(JSContextRef context, JSObjectRef, JSObjectRef thisObject, size_t argumentCount, const JSValueRef arguments[], JSValueRef* exception)
+JSValueRef JSIPC::getConnectionForProcessTarget(JSContextRef context, JSObjectRef, JSObjectRef thisObject, size_t argumentCount, const JSValueRef arguments[], JSValueRef* exception)
 {
+    RefPtr self = toWrapped(context, thisObject);
```
I don't think we should do this variable rename.

rniwa commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -1770,6 +1864,7 @@ void JSIPC::finalize(JSObjectRef object)
 const JSStaticFunction* JSIPC::staticFunctions()
 {
     static const JSStaticFunction functions[] = {
+        { "getConnectionForProcessTarget", getConnectionForProcessTarget, kJSPropertyAttributeDontDelete | kJSPropertyAttributeReadOnly },
```
We typically don't prefix a method with "get" prefix unless it has an out argument.
So I think we should call this `connectionForProcessTarget` instead.

rniwa commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -1805,59 +1900,69 @@ const JSStaticValue* JSIPC::staticValues()
-    JSC::JSLockHolder lock(globalObject->vm());
-    RefPtr jsIPC = toWrapped(context, thisObject);
-    if (!jsIPC) {
+    RefPtr self = toWrapped(context, thisObject);
```
I don't think we should do this rename.

rniwa commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -1036,92 +1111,90 @@ static std::optional<IPCStreamMessageInfo> extractIPCStreamMessageInfo(JSContext
 
+JSValueRef JSIPCStreamClientConnection::sendSyncMessage(JSContextRef context, JSObjectRef, JSObjectRef thisObject, size_t argumentCount, const JSValueRef arguments[], JSValueRef* exception)
+{
+    RefPtr self = toWrapped(context, thisObject);
```
I don't think we should do this rename.

rniwa commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -547,31 +495,73 @@ static std::optional<SyncIPCMessageInfo> extractSyncIPCMessageInfo(JSContextRef
+    JSGlobalContextRetain(JSContextGetGlobalContext(context));
+    JSValueProtect(context, callback);
+    IPC::Connection::AsyncReplyHandler handler {
+        [messageName, context, callback](IPC::Decoder* replyDecoder) {
```
I would have written as `handler = ` instead unless that doesn't work.

rniwa commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -547,31 +495,73 @@ static std::optional<SyncIPCMessageInfo> extractSyncIPCMessageInfo(JSContextRef
-    auto [destinationID, messageName, timeout] = *info;
+    auto encoder = makeUniqueRef<IPC::Encoder>(messageName, destinationID);
+    if (messageArguments) {
+        if (!encodeArgument(encoder.get(), context, messageArguments, exception))
```
Why not `messageArguments && !encodeArgument(~)`

rniwa commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -547,31 +495,73 @@ static std::optional<SyncIPCMessageInfo> extractSyncIPCMessageInfo(JSContextRef
+{
+    auto encoder = makeUniqueRef<IPC::Encoder>(messageName, destinationID);
+    if (messageArguments) {
+        if (!encodeArgument(encoder.get(), context, messageArguments, exception))
```
Ditto.

rniwa commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -842,7 +843,52 @@ JSValueRef JSIPCConnection::sendMessage(JSContextRef context, JSObjectRef, JSObj
+
+JSValueRef JSIPCConnection::sendWithAsyncReply(JSContextRef context, JSObjectRef, JSObjectRef thisObject, size_t argumentCount, const JSValueRef arguments[], JSValueRef* exception)
+{
+    RefPtr self = toWrapped(context, thisObject);
```
I prefer keeping the other name jsIPC, or maybe impl.

rniwa commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -870,7 +921,29 @@ JSValueRef JSIPCConnection::waitForMessage(JSContextRef context, JSObjectRef, JS
+
+JSValueRef JSIPCConnection::waitForAsyncReplyAndDispatchImmediately(JSContextRef context, JSObjectRef, JSObjectRef thisObject, size_t argumentCount, const JSValueRef arguments[], JSValueRef* exception)
+{
+    RefPtr self = toWrapped(context, thisObject);
```
I would have preferred to call this jsAPI or impl.

rniwa commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -1036,92 +1111,90 @@ static std::optional<IPCStreamMessageInfo> extractIPCStreamMessageInfo(JSContext
-    auto returnValue = JSValueMakeUndefined(context);
-    RefPtr jsStreamConnection = toWrapped(context, thisObject);
-    if (!jsStreamConnection) {
+    RefPtr self = toWrapped(context, thisObject);
```
Ditto to call this jsIPC or impl.

rniwa commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -1190,8 +1263,29 @@ JSValueRef JSIPCStreamClientConnection::waitForMessage(JSContextRef context, JSO
+
+JSValueRef JSIPCStreamClientConnection::waitForAsyncReplyAndDispatchImmediately(JSContextRef context, JSObjectRef, JSObjectRef thisObject, size_t argumentCount, const JSValueRef arguments[], JSValueRef* exception)
+{
+    RefPtr self = toWrapped(context, thisObject);
```
I would have called this jsIPC or impl.

rniwa commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -1805,59 +1900,69 @@ const JSStaticValue* JSIPC::staticValues()
-        return WebProcess::singleton().parentProcessConnection();
+    if (name == processTargetNameUI) {
+        RefPtr connection = WebProcess::singleton().parentProcessConnection();
+        if (!m_uiConnection || m_uiConnection->connection().ptr() != connection.get())
```
Nit: get() here seems unnecessary?

kkinnunen-apple committed:
IPC Testing API: asynchronous stream messages with replies not working
https://bugs.webkit.org/show_bug.cgi?id=272148
rdar://125084127

Reviewed by Ryosuke Niwa.

Implement support for asynchronous replies in JS testing API:
 - JSIPCConnection::sendWithAsyncReply
 - JSIPCConnection::waitForAsyncReplyAndDispatchImmediately
 - JSIPCStreamClientConnection::sendWithAsyncReply
 - JSIPCStreamClientConnection::waitForAsyncReplyAndDispatchImmediately

Previously in JS, the JSIPC object would support sending the async message
and receiving it as a promise. Stream connections would not support it.

Replace it with more consistent connection-specific sendWithAsyncReply
calls that take a callback parameter, similar to C++. Implement the
promise variant in JS side.

This way the async reply id is available for the tests, if they need it.

Organize the code so that the JSIPCConnection and
JSIPCStreamClientConnection share the common implementation in jsSend,
jsSendSync, jsSendWithAsyncReply, jsWaitForMessage,
jsWaitForAsyncReplyAndDispatchImmediately. Organize
JSIPCStreamClientConnection::prepareToSendOutOfStreamMessage to just
prepare the stream connection so that the above common functionality can
then be shared.

* LayoutTests/resources/ipc.js:
(connectionSendWithPromisedReply):
(IPC.sendWithPromisedReply):
(asyncFlush):
* Source/WebKit/Shared/IPCStreamTester.cpp:
(WebKit::IPCStreamTester::asyncPing):
(WebKit::IPCStreamTester::asyncMessage): Deleted.
* Source/WebKit/Shared/IPCStreamTester.h:
* Source/WebKit/Shared/IPCStreamTester.messages.in:
* Source/WebKit/Shared/IPCTester.cpp:
(WebKit::IPCTester::asyncPing):
* Source/WebKit/Shared/IPCTester.h:
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::JSIPCConnection::create):
(WebKit::IPCTestingAPI::JSIPCConnection::connection const):
(WebKit::IPCTestingAPI::JSIPCConnection::JSIPCConnection):
(WebKit::IPCTestingAPI::createErrorFromIPCError):
(WebKit::IPCTestingAPI::extractSyncIPCMessageInfo):
(WebKit::IPCTestingAPI::jsSend):
(WebKit::IPCTestingAPI::jsSendWithAsyncReply):
(WebKit::IPCTestingAPI::jsSendSync):
(WebKit::IPCTestingAPI::jsWaitForMessage):
(WebKit::IPCTestingAPI::jsWaitForAsyncReplyAndDispatchImmediately):
(WebKit::IPCTestingAPI::JSIPCConnection::staticFunctions):
(WebKit::IPCTestingAPI::JSIPCConnection::sendMessage):
(WebKit::IPCTestingAPI::JSIPCConnection::sendWithAsyncReply):
(WebKit::IPCTestingAPI::JSIPCConnection::sendSyncMessage):
(WebKit::IPCTestingAPI::JSIPCConnection::waitForMessage):
(WebKit::IPCTestingAPI::JSIPCConnection::waitForAsyncReplyAndDispatchImmediately):
(WebKit::IPCTestingAPI::JSIPCStreamClientConnection::staticFunctions):
(WebKit::IPCTestingAPI::extractIPCStreamMessageInfo):
(WebKit::IPCTestingAPI::JSIPCStreamClientConnection::prepareToSendOutOfStreamMessage):
(WebKit::IPCTestingAPI::JSIPCStreamClientConnection::sendMessage):
(WebKit::IPCTestingAPI::JSIPCStreamClientConnection::sendWithAsyncReply):
(WebKit::IPCTestingAPI::JSIPCStreamClientConnection::sendSyncMessage):
(WebKit::IPCTestingAPI::JSIPCStreamClientConnection::waitForMessage):
(WebKit::IPCTestingAPI::JSIPCStreamClientConnection::waitForAsyncReplyAndDispatchImmediately):
(WebKit::IPCTestingAPI::JSIPC::staticFunctions):
(WebKit::IPCTestingAPI::JSIPC::processTargetFromArgument):
(WebKit::IPCTestingAPI::JSIPC::addMessageListener):
(WebKit::IPCTestingAPI::messageNameFromArgument):
(WebKit::IPCTestingAPI::encodeArgument):
(WebKit::IPCTestingAPI::JSIPC::getConnectionForProcessTarget):
(WebKit::IPCTestingAPI::JSIPC::sendMessage):
(WebKit::IPCTestingAPI::JSIPC::waitForMessage):
(WebKit::IPCTestingAPI::JSIPC::sendSyncMessage):
(WebKit::IPCTestingAPI::sendMessageWithJSArguments): Deleted.
(WebKit::IPCTestingAPI::sendSyncMessageWithJSArguments): Deleted.
(WebKit::IPCTestingAPI::waitForMessageWithJSArguments): Deleted.
(WebKit::IPCTestingAPI::processTargetFromArgument): Deleted.
(WebKit::IPCTestingAPI::messageIDFromArgument): Deleted.
* LayoutTests/ipc/async-with-reply-expected.txt: Added.
* LayoutTests/ipc/async-with-reply.html: Added.
* LayoutTests/ipc/stream-async-with-reply-expected.txt: Added.
* LayoutTests/ipc/stream-async-with-reply.html: Added.

Canonical link: https://commits.webkit.org/277239@main
